hivemind/utils/grpc.py

```diff
@@ -6,19 +6,57 @@
-        size=array.shape,
-        dtype=array.dtype.name,
-        requires_grad=tensor.requires_grad)
+def serialize_torch_tensor(tensor: torch.Tensor, compression_type=CompressionType.NONE, allow_inplace=False) -> runtime_pb2.Tensor:
+    if compression_type == CompressionType.MEANSTD_LAST_AXIS_FLOAT16:
+        assert tensor.dtype == torch.float32
+        FP16_MAX = 65_504
+
+        tensor = tensor if allow_inplace else tensor.clone()
+        means = torch.mean(tensor, dim=-1, keepdim=True)
+        tensor.sub_(means)
+
+        stds = torch.square(tensor).sum(dim=-1, keepdim=True).div_(tensor.shape[-1]).sqrt_()
+        tensor.div_(stds)
+        tensor = tensor.clamp_(-FP16_MAX, FP16_MAX).to(torch.float16)
+
+        data = tensor.numpy().tobytes() + means.numpy().tobytes() + stds.numpy().tobytes()
+
+        proto = runtime_pb2.Tensor(
+            compression=compression_type,
+            buffer=data,
+            size=tensor.shape,
+            dtype='compressed_float16',
+            requires_grad=tensor.requires_grad)
+    else:
+        array = tensor.numpy()
+        proto = runtime_pb2.Tensor(
+            compression=compression_type,
+            buffer=array.tobytes(),
+            size=array.shape,
+            dtype=array.dtype.name,
+            requires_grad=tensor.requires_grad)
+
     return proto
 
 
-def deserialize_torch_tensor(tensor: runtime_pb2.Tensor) -> torch.Tensor:
-    # TODO avoid copying the array (need to silence pytorch warning, because array is not writable)
-    array = np.frombuffer(tensor.buffer, dtype=np.dtype(tensor.dtype)).copy()
-    return torch.as_tensor(array).view(tuple(tensor.size)).requires_grad_(tensor.requires_grad)
+def deserialize_torch_tensor(compressed_tensor: runtime_pb2.Tensor) -> torch.Tensor:
+    # TODO avoid copying the array (need to silence pytorch warning,x because array is not writable)
+    if compressed_tensor.compression == CompressionType.NONE:
+        array = np.frombuffer(compressed_tensor.buffer, dtype=np.dtype(compressed_tensor.dtype)).copy()
+        array = array.reshape(tuple(compressed_tensor.size))
+        tensor = torch.as_tensor(array).requires_grad_(compressed_tensor.requires_grad)
+    elif compressed_tensor.compression == CompressionType.MEANSTD_LAST_AXIS_FLOAT16:
+        stats_size = compressed_tensor.size
+        stats_size[-1] = 1
+        stats_count = 1
+        for i in stats_size:
+            stats_count *= i
+        means, stds = compressed_tensor.buffer[-8*stats_count:-4*stats_count], compressed_tensor.buffer[-4*stats_count:]
+        means = torch.as_tensor(np.frombuffer(means, dtype=np.float32)).view(*stats_size)
+        stds = torch.as_tensor(np.frombuffer(stds, dtype=np.float32)).view(*stats_size)
+        array = np.frombuffer(compressed_tensor.buffer[:-8 * stats_count], dtype=np.float16)
+        tensor = torch.as_tensor(array).to(torch.float32).view(*compressed_tensor.size).mul_(stds).add_(means)
+    return tensor
```

Comment:
```suggestion
    else:
        raise ValueError(f"Unknown compression type: {compressed_tensor.compression}")
    return tensor
```

rationale: if you received a compression type that you do not understand, its better to raise explicit error than "Undefined variable tensor"